src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleCloudStorageDAO.scala

```diff
@@ -1,81 +1,245 @@
+    val directory = getGroupDirectory(credential)
+
+    // we use a stream because we only care about the first access level that's valid for the user (checked in descending order)
+    val accessStream = Stream(groupAccessLevelsAscending.reverse: _*) map { accessLevel: WorkspaceAccessLevel =>
+      Try {
+        directory.members().get(makeGroupId(workspaceName, accessLevel), userId).execute()
+        accessLevel
+      }
+    }
+
+    // while the head of the stream is a failure, drop it and construct a new stream
+    accessStream.dropWhile(_.isFailure).headOption match {
+      case Some(Success(accessLevel)) => accessLevel
+      case None => WorkspaceAccessLevel.NoAccess
+    }
   }
 
-  override def putACL(userId: String, bucketName: String, acl: String): Unit = {
-    import system.dispatcher
-    val credential = getCredential(userId)
-    val pipeline = addHeader("Authorization",s"Bearer ${credential.getAccessToken}") ~> sendReceive
-    Await.result(pipeline(Put(s"https://www.googleapis.com/storage/v1/b/${bucketName}/acl",acl)),Duration.Inf)
+  override def getACL(bucketName: String, workspaceName: WorkspaceName): WorkspaceACL = {
+    val credential = getGroupServiceAccountCredential
+    val directory = getGroupDirectory(credential)
+
+    // return the maximum access level for each user
+    val aclMap = groupAccessLevelsAscending map { level: WorkspaceAccessLevel =>
+      val membersQuery = directory.members.list(makeGroupId(workspaceName, level)).execute()
+      // NB: if there are no members in this group, Google returns null instead of an empty list. fix that here.
+      val membersList = Option(membersQuery.getMembers).getOrElse(List.empty[Member].asJava)
+      membersList.map(_.getEmail -> level).toMap[String, WorkspaceAccessLevel]
+    } reduceRight(_ ++ _) // note that the reduction must go left-to-right so that higher access levels override lower ones
+
+    WorkspaceACL(aclMap)
+  }
+
+  /**
+   * Try to apply a list of ACL updates
+   *
+   * @return a map from userId to error message (an empty map means everything was successful)
+   */
+  override def updateACL(bucketName: String, workspaceName: WorkspaceName, aclUpdates: Seq[WorkspaceACLUpdate]): Map[String, String] = {
+    val credential = getGroupServiceAccountCredential
+    val directory = getGroupDirectory(credential)
+
+    aclUpdates map {
+      case WorkspaceACLUpdate(targetUserId, targetAccessLevel) => {
+        targetUserId -> updateUserAccessOrBarf(targetUserId, targetAccessLevel, workspaceName, directory)
+      }
+    } collect { case (user, Some(error)) => (user, error) } toMap
+  }
+
+  private def updateUserAccessOrBarf(targetUserId: String, targetAccessLevel: WorkspaceAccessLevel, workspaceName: WorkspaceName, directory: Directory): Option[String] = {
+    val currentAccessLevel = getMaximumAccessLevel(targetUserId, workspaceName)
+    if (currentAccessLevel != targetAccessLevel) {
+      Try {
```

Comment:
I will not hold up the review for this but I feel like there is a more idiomatic way to do this with flatmap or recoverwith or something... this code may as well be using lowercase try.